File: 1117-building-h2o/1117-building-h2o.cpp
class H2O {
public:
    mutex m1,m2;
    int h;
    H2O() {
        h=0;
        m2.lock();
    }

    void hydrogen(function<void()> releaseHydrogen) {
        m1.lock();
        h++;
        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();
        if(h==2)
        {
            h=0;
            m2.unlock();
        }
        else
        {
            m1.unlock();
        }
    }

    void oxygen(function<void()> releaseOxygen) {
        m2.lock();
        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        m1.unlock();
    }
};

File: 1117-building-h2o/1117-building-h2o.py
from threading import Lock
class H2O:
    def __init__(self):
        self.l1=Lock()
        self.l2=Lock()
        self.h=0
        self.l2.acquire()


    def hydrogen(self, releaseHydrogen: 'Callable[[], None]') -> None:
        self.l1.acquire()
        # releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen()
        self.h+=1
        if self.h==2:
            self.h=0
            self.l2.release()
        else:
            self.l1.release()

    def oxygen(self, releaseOxygen: 'Callable[[], None]') -> None:
        self.l2.acquire()
        # releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen()
        self.l1.release()